File: LinkedList/RemoveDuplicates2/remove.py
from typing import Optional
# Definition for singly-linked list.
class ListNode:
    def __init__(self, val=0, next=None):
        self.val = val
        self.next = next
class Solution:
    def deleteDuplicates(self, head: Optional[ListNode]) -> Optional[ListNode]:
        if head == None or head.next == None: return head
        sortMaker = None
        temp = head
        while temp and temp.next:
            if temp.val == temp.next.val:
                value = temp.val
                while temp and temp.val == value:
                    temp = temp.next
            else:
                if not sortMaker:
                    sortMaker = temp
                    head = sortMaker
                else:
                    sortMaker.next = temp
                    sortMaker = sortMaker.next
                if temp: temp = temp.next
        if temp and sortMaker and sortMaker.val != temp.val: sortMaker.next = temp
        else:
            if not sortMaker and temp: return temp
            elif sortMaker: sortMaker.next = None
            else: return  None
        return head

File: LinkedList/RemoveDuplicates2/remove.cpp
#include <iostream>

struct ListNode {
    int val;
    ListNode *next;
    ListNode() : val(0), next(nullptr) {}
    ListNode(int x) : val(x), next(nullptr) {}
    ListNode(int x, ListNode *next) : val(x), next(next) {}
};

class Solution {
public:
    ListNode* deleteDuplicates(ListNode* head) {
        if(head == NULL || head->next == NULL)return head;
        ListNode* sortMaker = NULL;
        ListNode* temp = head;
        while(temp && temp->next){
            if(temp->val == temp->next->val){
                int value = temp->val;
                while(temp && temp->val == value){
                    temp = temp->next;
                }
            }else{
                if(!sortMaker){
                    sortMaker = temp;
                    head = sortMaker;
                }
                else{
                    sortMaker->next = temp;
                    sortMaker = sortMaker->next;
                }
                if(temp)temp = temp->next;
            }
        }
        if(temp && sortMaker && sortMaker->val != temp->val)sortMaker->next = temp;
        else{ 
            if(!sortMaker && temp)return temp;
            else if (sortMaker) sortMaker->next = NULL;
            else return NULL;
        }
        return head;
    }
};